### DISCUSSION TITLE
Using Typescript on Client

### DISCUSSION THREAD

**Discussion by pavish** [mathesar-foundation/mathesar#145] [Maintainer discussions]
### Context

Currently we use Svelte with Javascript for the UI. I was thinking it would make more sense to use Svelte with Typescript for the following reasons:

* Typescript is a superset of JS, with type definitions. It compiles to JS during the same compilation step as Svelte.
* Having TS will save us time on documenting the types explicitly in JS code, and it ensures type safety. So we can avoid bugs early on.
* For eg., JS allows Error object to be constructed with anything, even though it treats the argument as a string. In our code, we had something like `new Error({ code: '503', detail: 'service unavailable' })`. While JS treats this as completely valid code, and during runtime will print `new error: [Object object]`, TS will show an error during compilation time.
* TS has excelled IDE support, and we can catch all of these during development itself.
* Regular JS code without types will work completely fine with a TS setup.

The major reason to use TS is that we will be accepting contributions from the community.

Using JS would mean that during review, we need to be extra careful with types, especially in cases where variables may belong to multiple types or be null or undefined, leading to subtle bugs which we won't be able to catch early on.

As the code base becomes larger with more contributors, this will become extremely hard to maintain. Using TS, will save us from that trouble.

### Drawbacks:
* Extra steps during linting and build process (But has no impact on build size)

### What is done
I have ported the existing code to TS in my current working branch: [readonly_table](https://github.com/centerofci/mathesar/tree/readonly_table)

Since we are in the early stages, the changes specific to the port are minimal. The existing code structure is the same, we just add type definitions to variables and methods.

### Some discussions/articles:
* https://news.ycombinator.com/item?id=26314756 [How an Anti-TypeScript “JavaScript developer” like me became a TypeScript fan]
* https://news.ycombinator.com/item?id=19597359 [When to Use TypeScript – A Detailed Guide Through Common Scenarios]
* https://news.ycombinator.com/item?id=22055341 [Problems with TypeScript]
* https://www.imaginarycloud.com/blog/typescript-vs-javascrip/ [Typescript vs Javascript]
* https://news.ycombinator.com/item?id=25043541 [Insights from Adopting TypeScript at Scale - On the server]

**Comment by kgodey** [comment]
@pavish I realized I never responded to this discussion. I'm fine with using TypeScript.

Can you add this decision to the engineering decisions wiki page?
